### PR TITLE
eglot-register-capability: Filter events per FileSystemWatcher.kind

### PR DESCRIPTION
**Pull request by leungbk**
* eglot.el (eglot-register-capability): Only send notifications of
interest, as determined by the optional FileSystemWatcher.kind
bitmask.

* eglot.el (eglot--file-change-types): Add new defconst.

When the FileSystemWatcher.kind property is omitted, use the default
value of 7, which is computed from taking the bitwise OR operation
WatchKind.Create (1) | WatchKind.Change (2) | WatchKind.Delete (4).

I learned of this problem in https://github.com/emacs-lsp/lsp-mode/issues/938, and submitted a fix in https://github.com/emacs-lsp/lsp-mode/pull/2085.

[Notwithstanding a hiccup related to my lack of knowledge of the protocol](https://github.com/emacs-lsp/lsp-mode/pull/2085#discussion_r470934392), I wrote that other PR on my own. The prior tests in `lsp-mode` helped me to write only the tests I submitted; they did not help me with the implementation of the actual bug fix. So I think I am able to assign copyright for this to the FSF.

Although I wrote tests in my other PR, I'm not sure how to write tests for this PR using Eglot's utilities. I would appreciate some help.

### CODE REVIEW

File: eglot.el [hunk 1]
```diff
@@ -244,6 +244,10 @@ let the buffer grow forever."
 
 (defconst eglot--{} (make-hash-table) "The empty JSON object.")
 
+(defconst eglot--file-change-types
```
Comment (joaotavora) [inline eglot.el:247]: We don't need a constant if it's only used once.  By itself the constant is meaningless, and when reading code I don't like to jump around.  I only use these things when they save on repetition.

File: eglot.el [hunk 2]
```diff
@@ -2611,33 +2615,35 @@ at point.  With prefix argument, prompt for ACTION-KIND."
...
          (globs (mapcar
-                 (eglot--lambda ((FileSystemWatcher) globPattern)
-                   (cons
+                 (eglot--lambda ((FileSystemWatcher) globPattern kind)
+                   (list
                     (eglot--glob-compile globPattern t t)
                     (eglot--glob-compile
-                     (replace-regexp-in-string "/[^/]*$" "/" globPattern) t t)))
+                     (replace-regexp-in-string "/[^/]*$" "/" globPattern) t t)
+                    (or kind
```
Comment (joaotavora) [inline eglot.el:2623]: same line please

File: eglot.el [hunk 3]
```diff
@@ -2611,33 +2615,35 @@ at point.  With prefix argument, prompt for ACTION-KIND."
...
-          (pcase-let ((`(,desc ,action ,file ,file1) event))
+          (pcase-let* ((`(,desc ,action ,file ,file1) event)
+                       (action-type (alist-get action eglot--file-change-types))
+                       (action-bit (ash 1 (1- action-type))))
             (cond
              ((and (memq action '(created changed deleted))
-                   (cl-find file (mapcar #'car globs)
-                            :test (lambda (f glob)
-                                    (funcall glob f))))
+                   (seq-some (pcase-lambda (`(,glob _ ,kind-mask))
```
Comment (joaotavora) [inline eglot.el:2639]: use `cl-some` please.  I'm not a big fan of `seq.el`.  Even better with less lines and commas.
```
(cl-loop for (glob _ kind-mask) in globs
            thereis (and (> (logand kind-mask action-bit) 0)
                                    (funcall glob file)))

**Review by joaotavora — CHANGES_REQUESTED** [review]
Thanks a lot, I meant do do this myself, but your work is appreciated. Please do these changes.

**Comment by joaotavora** [comment]
I wouldn't worry about copyright stuff in this case, *as long as you wrote this code yourself*.  The problem-solving techniques that are in your mind are in your mind, and no-one owns them.

> Although I wrote tests in my other PR, I'm not sure how to write tests for this PR using Eglot's utilities. I would appreciate some help.

To write an actual useful test for this, you have to try with an actual server.  See the existing test for the 'didChangeWatchFiles` thing, and maybe try to adapt it to `gopls` (i'd happily get rid of the `rls` dependency in the tests).

**Comment by leungbk** [comment]
> I wouldn't worry about copyright stuff in this case, _as long as you wrote this code yourself_. The problem-solving techniques that are in your mind are in your mind, and no-one owns them.

I wrote that `lsp-mode` patch "myself", or at least I think I did. @muirdm [alerted me of a small mistake](https://github.com/emacs-lsp/lsp-mode/pull/2085#discussion_r470930796) I made at the protocol level. Not sure if that's disqualifying for me.

I wrote this code myself, receiving no feedback from anyone outside this thread, and without consulting my other PR. Muir's review from before benefited this patch, but I think I would have spotted the problem myself eventually.

> To write an actual useful test for this, you have to try with an actual server. See the existing test for the 'didChangeWatchFiles`thing, and maybe try to adapt it to`gopls`(i'd happily get rid of the`rls` dependency in the tests).

I noticed servers prescribe their own parameters (namely, the `:globPatterns` and the `:kind` bitmask) for file-watchers. I wasn't sure how to force a server to use a `:kind` bitmask of our own choosing, and I recalled seeing someone register something like a file-watcher in one of `lsp-mode`'s files, so I looked at the relevant section and saw what they did in order to figure out how to use our own `:kind`. I think I can do something similar to what they did, but I'm not sure if using copying `lsp-mode`'s technique for use in a test in `eglot` would break the FSF's rules. Should I simply recuse from writing a test here? I'm not sure how to enforce a `:kind` without doing something similar to what I saw in `lsp-mode`.

I've informally tested this patch by running my patched repo with the `haskell-language-server`, which uses `:kind == 5` (create and delete operations), and confirmed that we're sending notifications for create and delete operations, and not sending notifications for change operations. Notwithstanding possible regression tests, it should be ready for review.

**Comment by leungbk** [comment]
I pushed another commit to clean up my changes to the inner function `handle-event`. From my brief testing with `haskell-language-server`'s file-watchers, it seems to work, but I basically never use `cl-loop`, so I'm not completely sure it's correct. Feel free to disregard if it looks off.

**Comment by leungbk** [comment]
Wrt tests, the options that come to mind are

1. Manually set the `file-watches` for the current server, which sort of defeats the purpose using a server to test. This is basically what is done in lsp-mode.
2. Use in the event-handler the bitmask from the file-watcher in the `client/registerCapability` server request, and ensure that the notifications desired (undesired) events are sent (not sent). For example, in your `rls-watches-files` test, I would only `eglot--wait-for` only for events known through the bitmask to be of interest. What I dislike about this approach is that the test code I'd write would look more or less the same as the code I've already written implementing this filter.
3. Avoid writing any tests for this.

What do you suggest, @joaotavora?

**Comment by skangas** [comment]
I have reviewed the code (but didn't test it), and it looks good to me.

As for tests, if I understand correctly the idea is to set `file-watches` to some value, then ask the server for that value, and then verifying that we only send whatever the server says that it wants. Right? If so, this sounds okay to me. Perhaps we should try one or two variations of it. However, if you prefer option 2, I have no objections to that approach either. In any case, both option 1 and 2 sound better than not having any test at all.

So if it sounds good to you, I'd suggest adding a test using whichever approach you prefer, and then we can finally get this merged.

Thanks.

**Comment by leungbk** [comment]
Submitted the patch to mainline Emacs instead.

**Pull request closed without merging**